chapter2.adoc

```diff
@@ -1,214 +1,207 @@
+This chapter introduces Sdsec ISA extension, which enhances the Sdext of _The RISC-V Debug Specification_ cite:[dbgspec]. The Sdsec extension provides privilege based protection for debug operations and trigger behaviors cite:[dbgspec]. Furthermore, it constrains trace functionality cite:[etrace] according to RISC-V privilege levels. 
 
+[[sdsecextdbg]]
 === External Debug
 
-By default, the extension forbids the following debug operation in <<dbops>> unless they are explicitly granted to the external debugger.
-	
+Chapter 3 of _The RISC-V Debug Specification_ cite:[dbgspec] outlines all mandatory and optional debug operations. The operations listed below are affected by the Sdsec extension, other operations remain unaffected. In the context of this chapter, *debug operations* refer to those listed below.
+
 [[dbops]]
-.The debug operations affected by Sdsec extension
-* Entering Debug Mode                                                     
+.Debug operations affected by Sdsec extension
+* Halting the hart to enter Debug Mode
 * Executing Program buffer                                                
 * Serving abstract commands (Access Register, Quick Access, Access Memory)
 
-When the debug operations are not granted, all the above operations issued by Debug Module or trigger will be pending or dropped. The subsequent subsections detail how debug operations are granted. 
-
-[[dbgaccpriv]]
-==== Debug Access Privilege
+When the external debug is disallowed in running privilege level, the hart behaves as the following: 
 
-The *debug access privilege* is defined as the privilege with which abstract commands or instructions in program buffers access hardware resources such as registers and memory. This privilege operates independently of hart privilege levels and exclusively affects operations within Debug Mode. Memory and register access within Debug Mode are subject to the *debug access privilege*, with all hardware protections, including MMU, PMP, and PMA, checked against it. This privilege is represented by the `prv` and `v` fields in `dcsr`, and it is updated to reflect the hart privilege level upon entering Debug Mode. Each hart has a dedicated *debug access privilege* and it may vary from each other. The permissible privilege levels programmable to `dcsr` in Debug Mode are elaborated in subsequent sections.
+[[dbgdisallowed]]
+* The hart will not enter Debug Dode on halt request (haltreq), single step or halt group. Halt request will be pending until debug is allowed again  
+* Triggers (with `action`=1) will not match or fire 
+* Abstract commands without halting (Access Register and Quick Access) will be dropped and set `cmderr` to 6 
```

Comment:
The thinking is that the quick access ( as well as the abstract command without halting) is supposed to debug in a minimum intrusive way and the debugger might just want to issue it when a specific event occurs. It doesnt make much sense to return a result after an unpredictable time as the status is already changed. Instead, it is straightforward to return an error and inform the debugger the current status is unavailable.

Reply:
@bcstrongx we discussed this topic in last week's TG meeting, unfortunately you were not there. 

Quick access implicitly implies halt, so no problem to follow halt behavior @AoteJin we should update spec

The main concern about pending for abstract command is that (unlike halt) abstract command can hang debugger forever, so we recommend the hart to drop and set error (keep current proposal)

Reply:
Sorry I couldn't make the meeting, I was on vacation last week.  The problem with this approach is that I doubt most harts know anything about Quick Access (QA) commands.  I would expect QA to be implemented as the DM sending a haltreq, followed by an abstract command, followed by a resumereq.  The savings is in avoiding the debugger interaction, but to the hart QA vs separate debugger commands is indistinguishable.

Obviously the DM can't track the priv mode status of all harts, so it would have to simply disallow QA on hart x if mdbgen[x]=0.  That seems undesirable.

Reply:
The concern is that abstract commands can only be issued to a hart at a time and when the command is issued, it cannot be canceled. Thus, when the external debugger happen to issue Quick Access to a debug-disallowed hart, the abstract command interface is exclusively occupied by the pending command, and the DM appears to be hung until the DM is reset. 
In order to avoid such situation, it might be better to differentiate the halt in QA from haltreq and halt-on-reset and make QA to return error when debug is disallowed.

Reply:
I understand the motivation, I'm just not sure how to implement it.  Only the hart knows if it is in a debuggable mode, and only the DM knows when the command is QA or not.  I suppose the DM could return cmderr=6 and clear haltreq if the hart doesn't halt within some time threshold?  Or we could disallow QA on hart x when mdbgen[x]=0.

Reply:
I have an immature thinking that the DM could use two bit vector to signify halt request to hart, e.g. 0b01 will signify normal halt that will pending when debug is disallowed, while 0b11 represents halt of QA. Since hart could differentiate those two halt, the hart can choose to have the halt request pending or respond exception.

Reply:
Expanding the DM to hart interface for this is not desirable, especially given that QA is not required.  Some implementations have dozens to hundreds of harts, so that's a lot of new wires all over the chip.  I think any of the other options I suggested would be better.

Reply:
It may be that debugger documentation should indicate to users that QA is blocking, so only use it if you want to wait for it to complete.  Otherwise do halt, command, resume separately.  Then the hart doesn't need to do anything special, any haltreq (whether for QA or not) remains pending until the hart is in a debuggable mode.

Another option is for the debugger to not block while waiting for QA to complete.  Only if the user wants to execute another abstract command does the debugger need to poll on the abstractcs.busy bit.

Reply:
I am think the amount of wires might not be a problem, because the debugger has the interface to access GPR, whose wires are non-trivial for each hart. The halt/reset/keepalive operation could re-use the wires with some dedicate encodings, depending on the implementation.
I am a little bit concerned about timeout, because it could varies between implementations. It is tricky to set an reasonable upper bound for it. Limiting the usage of QA by mdbgen seems too strict....

Actually the abstract command is not blocking and the debugger has to poll on busy status. But it just serves one outstanding command.

Reply:
I think the concern is reasonable.. Looks like it is better to have a quick live sync to resolve this issue. @AoteJin  can you open a new issue?

Reply:
I filed issue #50, @bcstrongx probably we could close this PR first and move discussion to the issue?